umap/static/umap/js/umap.js

```diff
@@ -2033,7 +2033,10 @@ L.U.Map.include({
   getFacetKeys: function () {
     return (this.options.facetKey || '').split(',').reduce((acc, curr) => {
       const els = curr.split('|')
-      acc[els[0]] = els[1] || els[0]
+      acc[els[0]] = {
+        "label": els[1] || els[0],
+        "type":  els[2] || "checkbox"
```

Comment:
Default for an `<input>` element [looks to be](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types) `text`.

Reply:
I see your point and also have some ideas:

1. Stick to data format (date, time, datetime, number, string, enum, ...) and assume the required input field format (min/max fields for dates, times, numbers, text for fuzzy matching strings, checkboxes for enum).
2. Extend the functionality even further and have both data format and input field format. With reasonable defaults. This would allow even more scenarios (e.g. date with sliders instead of min/max fields, enum with radio instead of checkboxes).

For now, I went for option 1 because option 2 can be done as an extension of option 1 in another improvement PR.

I did not adjust the naming in `umap.forms.js` because there the word checkbox refers to the input field type, so it seems correct to me.

If your happy with that, please resolve this conversation.

Reply:
I had some more minutes and actually implemented option 2 directly. This also gives us more flexibility for refactoring the parseDateField function (see https://github.com/umap-project/umap/pull/1491#discussion_r1439922296).

On the fly I also restricted the allowed values for dataType and inputType as well as allowed radio to be a inputType for enum dataType. Radio works quite ok, however right now there is no possibility to "deselect" the radio and show all features again. I will look into that tomorrow.

Reply:
Thanks for exploring that right away! :)

I'm not convinced about adding two parameters, dataType and inputType. I think that in an ideal world, the data type comes from the data iself, or from the layer, but we don't need to set the data type at the moment of computing the facets.
I'll like to implement a sort of json-schema like in uMap at some point, but of course we are not there yet.
So in the meantime, I'd only keep one parameter in the facet definition, which is the input type, and I'd derive the data type from it until we get the type from elsewhere.
I'd suggest for a first step to support those input types:
- checkbox (default) => data is enum, so we only check for equality
- radio (nice suggestion) => same data check
- minmax => two input min and max, and we check for the value to be inside (including boundaries)
- date => two inputs start and end, and we have a date picker
- datetime => same, but the date picker includes the time (if the browser supports it…)

I'm also concerned about keeping this nice enhancement as simple as possible, so we can let users use it and we can better priorise what to do next from their feedback. Also, once we have implement some syntax and users start using it, it means we need to support it for long.

What do you think ? :)

Reply:
I like the idea of allowing JSON input to define facets. The | splitting is becoming rather cumbersome and also brings the disadvantage of having to define all previous parameters, even if the user does not care about it (e.g. have to define label, otherwise I cannot define type). The only reason, I did not do this straight away were concerns about backwards compatibility.

I also like the idea of getting the data type from somewhere else. Brainstorming here a little: In think the easiest approach would be for the user to be able to define one data type per feature property name across all layers. We cannot have different data types per layer because facet search acts across all layers. Maybe we could still implement the data type definition per feature property name per layer and handle deviating data types during the facet search (e.g. excluding incompatible data types, defining groups of data types that work together, etc.)

As mentioned in one of my other comments, I am also rather unhappy with the complexity of the parseDateField function. Let's reduce the complexity of this change even further by not trying to be clever when the data is parsed but rely on existing JS functionality. However, I would change the input type naming a little to make things more transparent.

- checkbox (default) => one checkbox input per unique value, check for equality, data is passed through toString() method, should work for everything
- radio => same as checkbox but with radio input
- number_minmax => two number inputs for min and max, check for the value to be inside (including boundaries), data is passed through parseFloat function, NaNs are considered a match (otherwise features would disappear if property is empty)
- date_minmax => same as number_minmax but with date inputs and data is passed through Date() constructor
- datetime_minmax => same as date_minmax but with datetime-local inputs

Sounds good?

Reply:
Yes, keeping the `|` splitting for this PR.